File: gui/menu/workspace/workspace_high_to_low.cpp
#include "workspace_high_to_low.h"


MenuWorkspaceHighToLow::MenuWorkspaceHighToLow(QWidget* parent) : 
    IodeSettingsDialog(parent)
{
    setupUi(this);

    wInputFile = new WrapperFileChooser(label_input_file->text(), *fileChooser_input_file, REQUIRED_FIELD, 
                                        FILE_VARIABLES, EXISTING_FILE);
    wLastObs = new WrapperQTextEdit(label_last_obsvervation->text(), *textEdit_last_observation, OPTIONAL_FIELD);  
    wMean = new WrapperQTextEdit(label_mean->text(), *textEdit_mean, OPTIONAL_FIELD);  
    wSum = new WrapperQTextEdit(label_sum->text(), *textEdit_sum, OPTIONAL_FIELD);  

    mapFields["InputFile"] = wInputFile;
    mapFields["LastObs"] = wLastObs;
    mapFields["Mean"] = wMean;
    mapFields["Sum"] = wSum;

    className = "MENU_WORKSPACE_HIGH_TO_LOW";
    loadSettings();
}

MenuWorkspaceHighToLow::~MenuWorkspaceHighToLow()
{
    delete wInputFile;
    delete wLastObs;
    delete wMean;
    delete wSum;
}

void MenuWorkspaceHighToLow::slot_high_to_low()
{
    try
    {
        std::string filepath = wInputFile->extractAndVerify().toStdString();
        std::string last_obs_series = wLastObs->extractAndVerify().toStdString();
        std::string mean_series = wMean->extractAndVerify().toStdString();
        std::string sum_series = wSum->extractAndVerify().toStdString();

        if(last_obs_series.empty() && mean_series.empty() && sum_series.empty()) 
            throw std::invalid_argument(std::string("Cannot execute command 'high to low'") +  
                                        std::string("all series last observation, mean and sum are empty!"));

        if (!last_obs_series.empty()) high_to_low(HTOL_LAST, filepath, last_obs_series);
        if (!mean_series.empty()) high_to_low(HTOL_MEAN, filepath, mean_series);
        if (!sum_series.empty()) high_to_low(HTOL_SUM, filepath, sum_series);
    }
    catch (const std::exception& e)
    {
        QMessageBox::critical(this, tr("ERROR"), tr(e.what()));
        return;
    }

    this->accept();
}

void MenuWorkspaceHighToLow::help()
{
	
	QDesktopServices::openUrl(url_manual);
}

File: gui/menu/workspace/workspace_high_to_low.py
from PySide6.QtCore import Slot
from PySide6.QtWidgets import QWidget, QMessageBox

from settings import MixinSettingsDialog
from util.widgets.file_chooser import EnumFileMode
from .ui_workspace_high_to_low import Ui_MenuWorkspaceHighToLow

from iode import IodeFileType, variables, HighToLowType


class MenuWorkspaceHighToLow(MixinSettingsDialog):
    def __init__(self, parent: QWidget=None):
        super().__init__(parent)
        self.ui = Ui_MenuWorkspaceHighToLow()
        self.ui.setupUi(self)
        self.prepare_settings(self.ui)

        self.ui.fileChooser_input_file.enum_file_type = IodeFileType.FILE_VARIABLES
        self.ui.fileChooser_input_file.enum_file_mode = EnumFileMode.EXISTING_FILE

    @Slot()
    def slot_high_to_low(self):
        try:
            filepath: str = self.ui.fileChooser_input_file.filepath
            last_obs_series: str = ";".join(self.ui.textEdit_last_observation.toPlainText().splitlines()).strip()
            mean_series: str = ";".join(self.ui.textEdit_mean.toPlainText().splitlines()).strip()
            sum_series: str = ";".join(self.ui.textEdit_sum.toPlainText().splitlines()).strip()

            if not len(filepath):
                raise ValueError("'input file' is empty")

            # Check if all series names are empty
            if not last_obs_series and not mean_series and not sum_series:
                raise ValueError("The series 'last observation', 'mean' and 'sum' are all empty")

            # Call high_to_low function with appropriate arguments
            if last_obs_series:
                variables.high_to_low(HighToLowType.LAST, filepath, last_obs_series)
            if mean_series:
                variables.high_to_low(HighToLowType.MEAN, filepath, mean_series)
            if sum_series:
                variables.high_to_low(HighToLowType.SUM, filepath, sum_series)

            self.accept()
        except Exception as e:
            QMessageBox.warning(self, "WARNING", "Failed to build series of lower periodicity.\n" + str(e))
